File: handball_season_scraper.py
#!/usr/bin/env python3
"""
Handball Season Data Scraper
Fetches complete season data from Atrium Sports API
"""

import time
from datetime import datetime
from typing import List, Dict, Any, Optional
import logging

from base_handball_scraper import BaseHandballScraper
from config import config

logger = logging.getLogger(__name__)

class HandballSeasonScraper(BaseHandballScraper):
    def __init__(self, base_url: str = None):
        if base_url is None:
            base_url = config.api.base_url
        super().__init__(base_url)
        self.all_fixtures = []
    
    def make_request_and_extract_fixtures(self, params: Dict[str, Any] = None) -> Optional[List[Dict]]:
        """Make API request with error handling and extract fixtures"""
        data = self.make_request(params)
        if data:
            return self.extract_fixtures(data)
        return None
    
    def explore_pagination(self) -> Dict[str, Any]:
        """Explore different pagination parameters"""
        pagination_params = [
            {'page': 1},
            {'offset': 0, 'limit': 50},
            {'round': 1},
            {'matchday': 1},
        ]

        results = {}
        for params in pagination_params:
            logger.info(f"Testing pagination with: {params}")
            fixtures = self.make_request_and_extract_fixtures(params)
            if fixtures:
                results[str(params)] = len(fixtures)
                self.add_delay()

        return results
    
    def fetch_by_rounds(self, max_rounds: int = 50) -> List[Dict]:
        """Try to fetch data by round numbers"""
        all_fixtures = []

        for round_num in range(1, max_rounds + 1):
            logger.info(f"Fetching round {round_num}")

            # Try different round parameter names
            for round_param in ['round', 'matchday', 'gameweek']:
                fixtures = self.make_request_and_extract_fixtures({round_param: round_num})
                if fixtures:
                    logger.info(f"Found {len(fixtures)} fixtures in round {round_num}")
                    all_fixtures.extend(fixtures)
                    break
                self.add_delay()
            else:
                # No data found for this round
                logger.info(f"No data found for round {round_num}")
                if round_num > 5:  # Stop if we haven't found data for several rounds
                    break

            self.add_delay()

        return all_fixtures
    
    def fetch_by_pagination(self, max_pages: int = 100) -> List[Dict]:
        """Try to fetch data using pagination"""
        all_fixtures = []

        for page in range(1, max_pages + 1):
            logger.info(f"Fetching page {page}")

            # Try different pagination parameter names
            for page_param in ['page', 'offset']:
                if page_param == 'offset':
                    params = {'offset': (page - 1) * 50, 'limit': 50}
                else:
                    params = {page_param: page}

                fixtures = self.make_request_and_extract_fixtures(params)
                if fixtures:
                    logger.info(f"Found {len(fixtures)} fixtures on page {page}")
                    all_fixtures.extend(fixtures)
                    break
                self.add_delay()
            else:
                # No data found for this page
                logger.info(f"No data found for page {page}")
                if page > 3:  # Stop if we haven't found data for several pages
                    break

            self.add_delay()

        return all_fixtures
    
    def remove_duplicates(self, fixtures: List[Dict]) -> List[Dict]:
        """Remove duplicate fixtures based on fixtureId"""
        unique_fixtures = {}
        for fixture in fixtures:
            fixture_id = fixture.get('fixtureId')
            if fixture_id and fixture_id not in unique_fixtures:
                unique_fixtures[fixture_id] = fixture
        return list(unique_fixtures.values())
    
    def scrape_data(self) -> List[Dict]:
        """Main method to fetch all season data - implements abstract method"""
        logger.info("Starting to fetch handball season data...")

        # First, get the basic data
        base_fixtures = self.make_request_and_extract_fixtures()
        if not base_fixtures:
            logger.error("Failed to fetch base data")
            return []

        logger.info(f"Base request returned {len(base_fixtures)} fixtures")
        all_fixtures = base_fixtures.copy()

        # Since we got all fixtures in the base request, we might not need pagination
        # But let's still try to explore if there are more
        logger.info("Exploring pagination options...")
        pagination_results = self.explore_pagination()
        logger.info(f"Pagination exploration results: {pagination_results}")

        # Try fetching by rounds (might give us different data)
        logger.info("Attempting to fetch by rounds...")
        round_fixtures = self.fetch_by_rounds()
        if round_fixtures:
            all_fixtures.extend(round_fixtures)

        # Try fetching by pagination (might give us different data)
        logger.info("Attempting to fetch by pagination...")
        page_fixtures = self.fetch_by_pagination()
        if page_fixtures:
            all_fixtures.extend(page_fixtures)

        # Remove duplicates based on fixtureId
        final_fixtures = self.remove_duplicates(all_fixtures)
        logger.info(f"Total unique fixtures found: {len(final_fixtures)}")

        # Process fixtures into standardized format
        processed_matches = []
        for fixture in final_fixtures:
            processed_match = self.process_fixture_data(fixture)
            processed_matches.append(processed_match)

        return processed_matches
    

def main():
    """Main execution function"""
    scraper = HandballSeasonScraper()
    
    # Use the base class run method for consistent execution
    filename = scraper.run()
    
    if filename:
        print(f"📁 Data saved to: {filename}")
    else:
        print("❌ No data was saved")

if __name__ == "__main__":
    main()

File: handball_season_scraper.js
#!/usr/bin/env node
/**
 * Handball Season Data Scraper (Node.js)
 * Fetches complete season data from Atrium Sports API
 */

const https = require('https');
const fs = require('fs').promises;
const { URL } = require('url');

class HandballSeasonScraper {
    constructor(baseUrl = 'https://eapi.web.prod.cloud.atriumsports.com/v1/embed/248/fixtures') {
        this.baseUrl = baseUrl;
        this.allFixtures = [];
        this.delay = 1000; // Delay between requests in milliseconds
    }

    /**
     * Make HTTP request with error handling and extract fixtures
     */
    async makeRequest(params = {}) {
        const defaultParams = { locale: 'en-EN', ...params };
        const url = new URL(this.baseUrl);

        Object.keys(defaultParams).forEach(key => {
            url.searchParams.append(key, defaultParams[key]);
        });

        console.log(`Making request: ${url.toString()}`);

        return new Promise((resolve, reject) => {
            const options = {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
                    'Accept': 'application/json',
                    'Accept-Language': 'en-EN,en;q=0.9'
                }
            };

            const req = https.get(url.toString(), options, (res) => {
                let data = '';

                res.on('data', (chunk) => {
                    data += chunk;
                });

                res.on('end', () => {
                    try {
                        const jsonData = JSON.parse(data);

                        // Extract fixtures from the correct location in the response
                        if (jsonData && jsonData.data && jsonData.data.fixtures) {
                            const fixtures = jsonData.data.fixtures;
                            console.log(`Found ${fixtures.length} fixtures in response`);
                            resolve(fixtures);
                        } else {
                            console.warn('No fixtures found in response');
                            resolve([]);
                        }
                    } catch (error) {
                        console.error('JSON parse error:', error);
                        resolve(null);
                    }
                });
            });

            req.on('error', (error) => {
                console.error('Request error:', error);
                resolve(null);
            });

            req.setTimeout(30000, () => {
                req.destroy();
                console.error('Request timeout');
                resolve(null);
            });
        });
    }

    /**
     * Sleep function for delays
     */
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    /**
     * Explore different pagination parameters
     */
    async explorePagination() {
        const paginationParams = [
            { page: 1 },
            { offset: 0, limit: 50 },
            { round: 1 },
            { matchday: 1 }
        ];

        const results = {};
        
        for (const params of paginationParams) {
            console.log(`Testing pagination with:`, params);
            const fixtures = await this.makeRequest(params);

            if (fixtures) {
                results[JSON.stringify(params)] = fixtures.length;
            }

            await this.sleep(this.delay);
        }

        return results;
    }

    /**
     * Try to fetch data by round numbers
     */
    async fetchByRounds(maxRounds = 50) {
        const allFixtures = [];
        
        for (let roundNum = 1; roundNum <= maxRounds; roundNum++) {
            console.log(`Fetching round ${roundNum}`);
            
            // Try different round parameter names
            const roundParams = ['round', 'matchday', 'gameweek'];
            let foundData = false;
            
            for (const roundParam of roundParams) {
                const fixtures = await this.makeRequest({ [roundParam]: roundNum });

                if (fixtures && fixtures.length > 0) {
                    console.log(`Found ${fixtures.length} fixtures in round ${roundNum}`);
                    allFixtures.push(...fixtures);
                    foundData = true;
                    break;
                }

                await this.sleep(this.delay);
            }
            
            if (!foundData) {
                console.log(`No data found for round ${roundNum}`);
                if (roundNum > 5) { // Stop if we haven't found data for several rounds
                    break;
                }
            }
            
            await this.sleep(this.delay);
        }
        
        return allFixtures;
    }

    /**
     * Try to fetch data using pagination
     */
    async fetchByPagination(maxPages = 100) {
        const allFixtures = [];
        
        for (let page = 1; page <= maxPages; page++) {
            console.log(`Fetching page ${page}`);
            
            // Try different pagination parameter names
            const pageParams = ['page', 'offset'];
            let foundData = false;
            
            for (const pageParam of pageParams) {
                let params;
                if (pageParam === 'offset') {
                    params = { offset: (page - 1) * 50, limit: 50 };
                } else {
                    params = { [pageParam]: page };
                }
                
                const fixtures = await this.makeRequest(params);

                if (fixtures && fixtures.length > 0) {
                    console.log(`Found ${fixtures.length} fixtures on page ${page}`);
                    allFixtures.push(...fixtures);
                    foundData = true;
                    break;
                }
                
                await this.sleep(this.delay);
            }
            
            if (!foundData) {
                console.log(`No data found for page ${page}`);
                if (page > 3) { // Stop if we haven't found data for several pages
                    break;
                }
            }
            
            await this.sleep(this.delay);
        }
        
        return allFixtures;
    }

    /**
     * Main method to fetch all season data
     */
    async fetchAllData() {
        console.log('Starting to fetch handball season data...');
        
        // First, get the basic data
        const baseFixtures = await this.makeRequest();
        if (!baseFixtures) {
            console.error('Failed to fetch base data');
            return [];
        }

        console.log(`Base request returned ${baseFixtures.length} fixtures`);
        const allFixtures = [...baseFixtures];
        
        // Try to get more data using different methods
        console.log('Exploring pagination options...');
        const paginationResults = await this.explorePagination();
        console.log('Pagination exploration results:', paginationResults);
        
        // Try fetching by rounds
        console.log('Attempting to fetch by rounds...');
        const roundFixtures = await this.fetchByRounds();
        if (roundFixtures.length > 0) {
            allFixtures.push(...roundFixtures);
        }
        
        // Try fetching by pagination
        console.log('Attempting to fetch by pagination...');
        const pageFixtures = await this.fetchByPagination();
        if (pageFixtures.length > 0) {
            allFixtures.push(...pageFixtures);
        }
        
        // Remove duplicates based on fixtureId
        const uniqueFixtures = {};
        allFixtures.forEach(fixture => {
            const fixtureId = fixture.fixtureId;
            if (fixtureId && !uniqueFixtures[fixtureId]) {
                uniqueFixtures[fixtureId] = fixture;
            }
        });
        
        const finalFixtures = Object.values(uniqueFixtures);
        console.log(`Total unique fixtures found: ${finalFixtures.length}`);
        
        return finalFixtures;
    }

    /**
     * Save fixtures data to JSON file
     */
    async saveData(fixtures, filename = null) {
        if (!filename) {
            const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, -5);
            filename = `handball_season_${timestamp}.json`;
        }
        
        const dataToSave = {
            metadata: {
                total_fixtures: fixtures.length,
                scraped_at: new Date().toISOString(),
                source_url: this.baseUrl
            },
            fixtures: fixtures
        };
        
        await fs.writeFile(filename, JSON.stringify(dataToSave, null, 2), 'utf8');
        console.log(`Data saved to ${filename}`);
        return filename;
    }
}

/**
 * Main execution function
 */
async function main() {
    const scraper = new HandballSeasonScraper();
    
    try {
        const fixtures = await scraper.fetchAllData();
        
        if (fixtures.length > 0) {
            const filename = await scraper.saveData(fixtures);
            console.log(`\n✅ Successfully scraped ${fixtures.length} fixtures`);
            console.log(`📁 Data saved to: ${filename}`);
            
            // Print some statistics
            if (fixtures.length > 0) {
                const dates = fixtures
                    .map(f => f.startTimeLocal)
                    .filter(date => date)
                    .sort();
                
                if (dates.length > 0) {
                    console.log(`📅 Date range: ${dates[0].slice(0, 10)} to ${dates[dates.length - 1].slice(0, 10)}`);
                }
                
                const teams = new Set();
                fixtures.forEach(fixture => {
                    if (fixture.competitors) {
                        fixture.competitors.forEach(competitor => {
                            if (competitor.name) {
                                teams.add(competitor.name);
                            }
                        });
                    }
                });
                console.log(`🏆 Teams found: ${teams.size}`);
            }
        } else {
            console.log('❌ No fixtures found');
        }
        
    } catch (error) {
        console.error('❌ Error occurred:', error);
    }
}

// Run if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

module.exports = HandballSeasonScraper;
